File: 451-sort-characters-by-frequency/sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char,int> mp;
        for(char ch: s) mp[ch]++;
        priority_queue<pair<int,char>> pq;
        for(auto &it: mp)
        {
            pq.push({it.second,it.first});
        }

        string ans="";
        while(!pq.empty())
        {
            int freq=pq.top().first;
            char ch=pq.top().second;
            pq.pop();
            while(freq--) ans+=ch;
        }
        return ans;
    }
};

// class Solution {
// public:
//     string frequencySort(string s) {
//         int n=s.length();
//         pair<int,char> freq(26,0);
//         for(int i=0;i<26;i++)
//         {
//             freq[i]={0,i+'a'};
//         }

//         for(char ch:s)
//         {
//             freq[ch-'a'].first++;
//         }
//         sort(freq.begin(),freq.end());

//         string ans="";
//         for(auto &it: freq)
//         {
//             int freq=it.first;
//             char ch=it.second;
//             if(freq==0) break;
//             while(freq--) ans+=ch;
//         }
//         return ans;
//     }
// };

File: 451-sort-characters-by-frequency/sort-characters-by-frequency.java
class Solution {
    class Pair {
        int freq;
        char ch;

        public Pair(int freq,char ch) {
            this.freq = freq;
            this.ch = ch;
        }
    }

    public String frequencySort(String s) {
        int n = s.length();
        Pair[] hash = new Pair[256];

        for(int i=0;i<256;i++) {
            hash[i] = new Pair(0,(char)(i));
        }

        for(char ch : s.toCharArray()) {
            hash[ch].freq++;
        }

        Arrays.sort(hash,(a,b) -> {
            if(a.freq != b.freq) return b.freq - a.freq;
            else return a.ch - b.ch;
        });

        StringBuilder sb = new StringBuilder();
        for(int i=0;i<256;i++) {
            int freq = hash[i].freq;
            if(freq == 0) break;
            char ch = hash[i].ch;
            while(freq > 0) {
                sb.append(ch);
                freq--;
            }
        }
        return sb.toString();
    }
}